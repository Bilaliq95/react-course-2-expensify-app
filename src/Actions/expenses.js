import uuid from 'uuid';
import database from '../firebase/firebase'
import expenses from '../tests/fixtures/expenses';


//Action Generators

export const addExpense=(expense)=>{
   
    return{
        type:'ADD_EXPENSE',
        expense
    }
}

export const startAddExpense=(expenseData={})=>{
return (dispatch,getState)=>{
    const uid=getState().auth.uid;
const {
    description='',
    note='',
    amount=0,
    createdAt=0

}=expenseData;
const expense={description,note,amount,createdAt};
return database.ref(`users/${uid}/expenses`).push(expense).then((ref)=>{  //This returns a promise
    dispatch(addExpense({
        id:ref.key,
        ...expense
    }));
})
}
}


export const removeExpense=({id})=>{
return{
    type:'REMOVE_EXPENSE',
    id
}
}

export const startRemoveExpense=({id})=>{
return (dispatch,getState)=>{
    const uid=getState().auth.uid;
database.ref(`users/${uid}/expenses/`+id).remove().then(()=>{
    dispatch(removeExpense({id}));
});
}
}

export const editExpense=(id,updates)=>{
    return{
    type:'EDIT_EXPENSE',
    id,
    updates
}
}

export const startEditExpense=(id,updates)=>{

return (dispatch,getState)=>{
    const uid=getState().auth.uid;
    database.ref(`users/${uid}/expenses/`+id).update(updates).then(()=>{
        dispatch(editExpense(id,updates));
    });
}
}

//SET EXPENSES
export const setExpenses=(expenses)=>{
    return {
        type:'SET_EXPENSES',
        expenses
    }
}

export const startSetExpenses=()=>{
    return(dispatch,getState)=>{
        const uid=getState().auth.uid;
      return  database.ref(`users/${uid}/expenses`).once('value').then((snapshot) => {
        const expenses = [];   
                snapshot.forEach((childSnapsot) => { 
                    expenses.push({
                        id: childSnapsot.key,
                        ...childSnapsot.val()
                    });

                }); 
                dispatch(setExpenses(expenses))
            });
            
    };
};