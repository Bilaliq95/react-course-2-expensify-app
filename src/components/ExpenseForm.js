import React from 'react';
import moment from 'moment'; 
import {SingleDatePicker} from 'react-dates';





 class ExpenseForm extends React.Component{


extractProps=()=>{

    if(this.props.editExpenseProp===undefined)
    {
            
        return{    
                description:'',
                 note:'',
                amount:'',
              createdAt:moment(),
               calendarFocused:false,
                error:undefined
         }
    }
        
    else
    {
         
        return {
           
               ...this.props.editExpenseProp,  
                 error:undefined,
               calendarFocused:false,
                createdAt:moment(this.props.editExpenseProp.createdAt),
                amount: (this.props.editExpenseProp.amount/100).toString()
            
            }

    }
}



state=this.extractProps();

    onDescriptionChange=(e)=>{
        const description=e.target.value;
        this.setState(()=>{
            return{
                description:description
            }
        })
    }
    onNoteChange=(e)=>{
        const note=e.target.value;
        this.setState(()=>{
            return{
                note:note
            }
        })
    }
    onAmountChange=(e)=>{
        const amount=e.target.value;
        if(!amount || amount.match(/^\d{1,}(\.\d{0,2})?$/))
        {
            this.setState(()=>{
                return {
                    amount:amount
                }
            })
        }
    }

    onDateChange=(createdAt)=>{
        
       if(createdAt){
        this.setState(()=>{
            return {
                createdAt:createdAt
        }})
    }
            
     }
    onFocusChange=({focused})=>{
        this.setState(()=>{
               return{
             calendarFocused:focused
            }
        })
    }
     
render(){
    return(
       
       
       
        <form className="form" onSubmit={(e)=>{
            e.preventDefault();
          if(!this.state.description || !this.state.amount)
          {
            this.setState(()=>{
                return{
                    error:'Amount and Description are mandatory fields.'
                }
            })
          }
          else{

            this.setState(()=>{
                return{
                    error:''
                }
            })

            this.props.onSubmit({
                description:this.state.description,
                amount:parseFloat(this.state.amount,10)*100,
                createdAt:this.state.createdAt.valueOf(),
                note:this.state.note
            })
            {

            }
          }
        }}>
        <p className="form__error">{this.state.error}</p>
        <input
        type='text'
        placeholder='Description'
        autoFocus
        className="text-input"
        value={this.state.description}
        onChange={this.onDescriptionChange }>
        </input>
        <input
        type='text'
        className="text-input"
        placeholder='Amount'
        value={this.state.amount}
        onChange={this.onAmountChange}
        ></input>
        <SingleDatePicker 
        date={this.state.createdAt}                               //where you want to start
        onDateChange={this.onDateChange} //called with above prop
        focused={this.state.calendarFocused}
        onFocusChange={this.onFocusChange}
        numberOfMonths={1}
        isOutsideRange={()=>false}
        
        />
        <textarea placeholder='Add a note for your expense(optional)'
        value={this.state.note}
        className="textarea"
        onChange={this.onNoteChange }>
        </textarea>
        <div>
        <button className="button" >Save Expense</button>
        </div>
        </form>
        

    )
}
}


export default ExpenseForm


//Why are we using state here???
//The idea here is to keep the track of what user enters into the boxes and do something meaningful with it as the user presses add expense button
//It is still unclear why do we need a state at all.