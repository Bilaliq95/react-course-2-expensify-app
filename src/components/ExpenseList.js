import React from 'react'
import {connect} from 'react-redux'
import ExpenseListItem from './ExpenseListItem'
import getVisibleExpenses from '../Selectors/expenses'




export const ExpenseList=(props)=>(
    <div>
    {
        props.expenses.length===0?(<p>No Expenses</p>):(

            props.expenses.map((expense)=>{
                return(
                  <ExpenseListItem key={expense.id} expense={expense}/>
            )
            })
        )
    }
     
   

    </div>
)

const mapStateToProps=(state)=>{
    return{
        expenses:getVisibleExpenses(state.expenses,state.filters) //We will use the same name we used in the store.
    }

}


export default connect(mapStateToProps)(ExpenseList)

 