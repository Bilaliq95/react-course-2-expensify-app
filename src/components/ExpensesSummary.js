
import React from 'react'
import {connect} from 'react-redux'
import getVisibleExpenses from '../Selectors/expenses'
import getExpensesTotal from '../Selectors/expenses-total'
import numeral from 'numeral'
import {Link} from 'react-router-dom';

const ExpensesSummary=(props)=>(
    <div className="page-header">
    <div className="content-container">
    {
    props.expenses.length>1?(<h1 className="page-header__title">{`Viewing ${props.expenses.length} expenses totalling ${numeral(getExpensesTotal(props.expenses)/100).format('$0,0.00')}`}</h1>):(
     <h1 className="page-header__title">{ `Viewing ${props.expenses.length} expense totalling ${numeral(getExpensesTotal(props.expenses)/100).format('$0,0.00')}` 
     }</h1>)    
    }
    <div className="page-header__actions">
    <Link className="button" to="/create">Add Expense</Link> 
    </div>
    </div>
    </div>
)

const mapStateToProps=(state)=>{
    return{
        expenses:getVisibleExpenses(state.expenses,state.filters) //We will use the same name we used in the store.
    }

}


export default connect(mapStateToProps)(ExpensesSummary)