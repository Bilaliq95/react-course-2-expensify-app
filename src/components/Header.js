import React from 'react';
import {NavLink} from 'react-router-dom';

const Header=()=>(
       
    <header>
    <h1>Expensify</h1>
    
    
    <NavLink activeClassName="active-class" to="/" exact={true}>Dashboard </NavLink> 
    <NavLink activeClassName="active-class" to ="/create">Create </NavLink>
    <NavLink activeClassName="active-class" to ="/edit">Edit </NavLink>
    <NavLink activeClassName="active-class" to="help">Help </NavLink>
    
    </header>
    
    
    
    )


    export default Header;