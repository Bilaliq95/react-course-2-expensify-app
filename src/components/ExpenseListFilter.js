import React from 'react';
import {connect} from 'react-redux';
import{setTextFilter} from '../Actions/filter'
import{SortByAmount,SortByDate,SetStartDate,SetEndDate} from '../Actions/filter'
import {DateRangePicker} from 'react-dates'



export class ExpenseListFilter extends React.Component{

state={
    calendarFocused:null    
};
onDatesChange=({startDate,endDate})=>{
this.props.SetStartDate(startDate)
this.props.SetEndDate(endDate)
};


onFocusChange=(calendarFocused)=>{
this.setState(()=>{
    return{
        calendarFocused:calendarFocused
    }
})
}

onTextChange=(e)=>{
    
this.props.setTextFilter(e.target.value);
    
}

onSortChange=(e)=>{
    if (e.target.value==='amount')
    {
        this.props.SortByAmount()   
    } 
    else {

            this.props.SortByDate()
    }
}

render(){
   return  (
        <div>
        <input type='text' value={this.props.filters.text} onChange={this.onTextChange}></input>
    
      
        <select value={this.props.filters.sortBy} onChange={this.onSortChange}>
        <option value='amount'>Amount</option>
        <option value='date'>Date</option>
        </select>
       <DateRangePicker startDate={this.props.filters.startDate}
       endDate={this.props.filters.endDate}
       onDatesChange={this.onDatesChange}
        focusedInput={this.state.calendarFocused}
        onFocusChange={this.onFocusChange}
        numberOfMonths={1}
        isOutsideRange={()=>false}
        showClearDates={true}
       />
        </div>
    )
}

}

const mapStateToProps=(state)=>{
return{
    filters:state.filters
    
}
}

const mapDispatchToProps=(dispatch)=>({
     setTextFilter:(text)=>(dispatch(setTextFilter(text))),
     SortByAmount:()=>(dispatch(SortByAmount())),
     SortByDate:()=>(dispatch(SortByDate())),
     SetStartDate:(startDate)=>(dispatch(SetStartDate(startDate))),
     SetEndDate:(endDate)=>(dispatch(SetEndDate(endDate)))
})

export default connect(mapStateToProps,mapDispatchToProps)(ExpenseListFilter)