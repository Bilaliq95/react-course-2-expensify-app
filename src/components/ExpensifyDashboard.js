import React from 'react';
import ExpenseList from './ExpenseList'
import ExpenseListFilter from './ExpenseListFilter'
import getExpensesTotal from '../Selectors/expenses-total'
import expenses from '../tests/fixtures/expenses'
import ExpensesSummary from '../components/ExpensesSummary'
const ExpensifyDashboard=()=>(
    <div>
    <ExpensesSummary/>
    <ExpenseListFilter/>    
     <ExpenseList/>
    </div>
    
    );

 export default ExpensifyDashboard;   