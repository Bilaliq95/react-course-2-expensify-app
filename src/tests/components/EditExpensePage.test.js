import React from 'react';
import {shallow} from 'enzyme';
import {EditExpensePage} from '../../components/EditExpensePage';
import expenses from '../fixtures/expenses'

let wrapper,history,editExpense,removeExpense

beforeEach(()=>{
    editExpense=jest.fn();
    removeExpense=jest.fn();
    history=  {push:jest.fn()}
    wrapper=shallow(<EditExpensePage  editExpenseProp={expenses[2]}  removeExpense={removeExpense} editExpense={editExpense} history={history}/>)
});

test('Should render correct edit Expense component',()=>{
    expect(wrapper).toMatchSnapshot();
})

test('Should handle edit expense',()=>{
wrapper.find('ExpenseForm').prop('onSubmit')(expenses[2])
expect (history.push).toHaveBeenLastCalledWith('/');
expect (editExpense).toHaveBeenLastCalledWith(expenses[2].id,expenses[2]);
    
})

test('Should handle remove expense',()=>{
    wrapper.find('button').simulate('click')
    expect (history.push).toHaveBeenLastCalledWith('/');
    expect (removeExpense).toHaveBeenLastCalledWith({id:expenses[2].id});
        
    });