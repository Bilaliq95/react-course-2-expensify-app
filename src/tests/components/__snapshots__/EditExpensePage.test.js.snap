// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Should render correct edit Expense component 1`] = `
<div>
  <ExpenseForm
    onSubmit={[Function]}
  />
  <button
    onClick={[Function]}
  >
    Remove Expense
  </button>
</div>
`;
